Add new --matrix option to multiply commands

cdrini opened the pull request:
Closes #517

gustavohenke commented on src/index.ts:
```diff
@@ -103,6 +103,11 @@ export type ConcurrentlyOptions = Omit<BaseConcurrentlyOptions, 'abortSignal' |
     additionalArguments?: string[];
+
+    /**
+     * This command should be run multiple times, for each of the provided matrices.
```
Can you please update this comment? It looks like it's assuming some context from elsewhere.

gustavohenke commented on src/concurrently.ts:
```diff
@@ -179,6 +185,10 @@ export function concurrently(
         new ExpandWildcard(),
     ];
 
+    if (options.matrices?.length) {
```
Looks like matrices have preference over passthrough arguments using the same `{number}` syntax. Was it intended?

It needs to be called out in the help and in the docs.

gustavohenke reviewed (changes requested):
Thanks for taking this up!
I made several suggestions and asked a few questions in this pass.

I'll update [the docs](https://github.com/open-cli-tools/concurrently/blob/main/docs/cli/passthrough-arguments.md) to mention matrices as well in a follow-up, unless you feel like doing it while keeping it on brand?

paescuj commented:
@cdrini Wanna take this up again?

cdrini commented:
Thank you for rebasing, @paescuj ! And thank you for the code review @gustavohenke . Apologies for the delay was a bit stuck on the right approach. Taking another look today.

cdrini commented on src/command-parser/expand-matrices.ts:
```diff
@@ -0,0 +1,62 @@
+    private _bindings: string[][];
+
+    constructor(private readonly matrices: readonly string[][]) {
+        this.matrices = matrices;
```
I think I cleaned all these up? I'm not too familiar with the best practices for TS for fields like these, let me know if there's something missing!

cdrini commented:
Ok! I updated this up with the comments. I decided to tweak the interface slightly to allow/enforce _named_ matrix dimensions, and a `M:` prefix for the matrix placeholder in the command (eg `--matrix "os:windows linux" 'echo {M:os}'`). This not only fixes the collisions issue with the placeholder for pass-through arguments, but also avoids an issue I've noticed with GNU parallel, which is that it very quickly becomes unreadable with very many eg `{1}` `{3}` `{2}`. It also avoids having to have extra options for names (something GNU parallel supports IIRC), which I think fragments the interface a bit. Now there's one interface that kind of handles it all. Although at the cost of being a bit more verbose.

I updated the description with the new interface. Would love to know your thoughts!

Note: I also fixed a naming error I'd made in the code; there aren't multiple `matrices`, there is just one `matrix`!

~Also if anyone has any thoughts on why that mac test is failing... 🤔~ Thank you to whoever I assume re-ran it.
